tmt/steps/__init__.py

```diff
@@ -863,42 +871,55 @@ def login(context: Any, **kwargs: Any) -> None:
 
     @classmethod
     def plugins(cls, step: Step) -> List['Login']:
+        from tmt.steps.execute import ExecutePlugin
         """ Return list of login instances for given step """
         if not Login._enabled:
             return []
+        if cls._opt('test') and step.name == 'execute':
+            for p in step.phases(classes=(ExecutePlugin,)):
+                p._login_after_test = Login(step, 90)
         return [Login(step, phase) for phase in cls.phases(step)]
 
     def go(self) -> None:
         """ Login to the guest(s) """
-        # Verify possible test result condition
-        count: Dict[str, int] = dict()
-        expected_results: Optional[List[str]] = self.opt('when')
 
-        assert isinstance(self.parent, Step)
-        assert hasattr(self.parent, 'plan') and self.parent.plan is not None
+        if self._enabled_by_results(self.parent.plan.execute.results()):
+            self._login()
 
+    def _enabled_by_results(self, results: List['tmt.base.Result']) -> bool:
+        """ Verify possible test result condition """
+        expected_results: Optional[List[str]] = self.opt('when')
         if expected_results:
-            for expected_result in expected_results:
-                count[expected_result] = len([
-                    result for result in self.parent.plan.execute.results()
-                    if result.result == expected_result])
-            if not any(count.values()):
+            count = Counter([r.result for r in results
+                             if r.result in expected_results])
+            if not len(count):
```

Comment:
IIUIC, the purpose of the test was to find if there's any `r.result` in `expected_results` (the number of such occurrences is ignored). So, would `if any(r.result in expected_results for r in results):` do? No intermediate counter and it answers the question as soon as possible - first `r.result` in `expected_results` satisfies the branch.

Reply:
Also, mypy should complain that `expected_results` is `Optional` yet used in `in` test. `assert expected_results is not None` or `expected_results: Optional[List[str]] = self.opt('when') or []` should be needed.

Reply:
> IIUIC, the purpose of the test was to find if there's any `r.result` in `expected_results` (the number of such occurrences is ignored). So, would `if any(r.result in expected_results for r in results):` do? No intermediate counter and it answers the question as soon as possible - first `r.result` in `expected_results` satisfies the branch.

There was the result counting, so I tried to refactor it in a more generic way. But you are right. It does not have any additional value to the code. I changed it to a similar for loop as you proposed. Fixed.

Reply:
> Also, mypy should complain that `expected_results` is `Optional` yet used in `in` test. `assert expected_results is not None` or `expected_results: Optional[List[str]] = self.opt('when') or []` should be needed.

It is not complaining because of the `        if expected_results:`, but I can add the assert if you want to.

Reply:
I see, you're absolutely correct. There's no need for extra assert, `if` is good enough.

Reply:
Ack, resolving the comment.